File: app/services/metadata_fetcher.rb
# Fetches TV series metadata and episode details from the TVMaze API.
#
# TVMaze is free, requires no API key, and returns posters, descriptions,
# genres, ratings, and per-episode details in a single request.
#
# API docs: https://www.tvmaze.com/api
# Rate limit: 20 calls per 10 seconds per IP.
#
# Usage:
#   MetadataFetcher.fetch!(series)            # fetch metadata for a Series record
#   MetadataFetcher.search("Breaking Bad")    # search without saving
#
class MetadataFetcher
  BASE_URL = 'https://api.tvmaze.com'.freeze

  class << self
    # Fetch metadata for a Series record and update it + its episodes.
    # Returns true if metadata was found and applied, false otherwise.
    def fetch!(series)
      data = search(series.name)
      return false unless data

      apply_series_metadata!(series, data)
      apply_episode_metadata!(series, data['_embedded']&.dig('episodes') || [])
      true
    rescue StandardError => e
      Rails.logger.warn("MetadataFetcher: failed for '#{series.name}' — #{e.message}")
      false
    end

    # Search TVMaze for a show by name. Returns the raw JSON hash
    # with embedded episodes, or nil if not found.
    def search(query)
      uri = URI("#{BASE_URL}/singlesearch/shows")
      uri.query = URI.encode_www_form(q: query, embed: 'episodes')

      response = http_get(uri)
      return nil unless response.is_a?(Net::HTTPSuccess)

      JSON.parse(response.body)
    rescue StandardError => e
      Rails.logger.warn("MetadataFetcher: search failed for '#{query}' — #{e.message}")
      nil
    end

    private

    def apply_series_metadata!(series, data)
      poster_url = data.dig('image', 'original') || data.dig('image', 'medium')
      summary = strip_html(data['summary'])

      series.update!(
        tvmaze_id: data['id'],
        poster_url: poster_url,
        description: summary,
        genres: Array(data['genres']).join(', '),
        rating: data.dig('rating', 'average'),
        premiered: data['premiered'],
        status: data['status'],
        imdb_id: data.dig('externals', 'imdb')
      )

      Rails.logger.info("MetadataFetcher: updated series '#{series.name}' (TVMaze ID: #{data['id']})")
    end

    def apply_episode_metadata!(series, api_episodes)
      return if api_episodes.empty?

      # Build a lookup by SxxExx code from the API data
      api_lookup = {}
      api_episodes.each do |ep|
        season = ep['season']
        number = ep['number']
        next unless season && number

        code = format('S%02dE%02d', season, number)
        api_lookup[code] = ep
      end

      # Match against our locally-scanned episodes
      matched = 0
      series.episodes.find_each do |episode|
        api_ep = api_lookup[episode.code]
        next unless api_ep

        attrs = {}
        summary = strip_html(api_ep['summary'])
        attrs[:description] = summary if summary.present?
        attrs[:air_date]    = api_ep['airdate'] if api_ep['airdate'].present?
        attrs[:runtime]     = api_ep['runtime'] if api_ep['runtime']
        attrs[:tvmaze_id]   = api_ep['id'] if api_ep['id']

        if attrs.any?
          episode.update!(attrs)
          matched += 1
        end
      end

      Rails.logger.info("MetadataFetcher: matched #{matched}/#{series.episodes.count} episodes with TVMaze data")
    end

    def strip_html(html)
      return nil if html.blank?

      # Remove HTML tags and decode common entities
      text = html.gsub(/<[^>]+>/, '')
      text = text.gsub('&amp;', '&')
                 .gsub('&lt;', '<')
                 .gsub('&gt;', '>')
                 .gsub('&quot;', '"')
                 .gsub('&#39;', "'")
                 .gsub('&nbsp;', ' ')
      text.strip.presence
    end

    def http_get(uri)
      require 'net/http'
      require 'json'
      require 'uri'

      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true
      http.open_timeout = 10
      http.read_timeout = 15

      request = Net::HTTP::Get.new(uri)
      request['Accept'] = 'application/json'

      http.request(request)
    end
  end
end

File: desktop/electron/services/metadata-fetcher.js
// Fetches TV show metadata from TVMaze API.
// No API key needed. Rate limit: 20 calls/10s.

const db = require('../db')

const BASE_URL = 'https://api.tvmaze.com'

function stripHtml(html) {
  if (!html) return null
  let text = html.replace(/<[^>]+>/g, '')
  text = text.replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&nbsp;/g, ' ')
  return text.trim() || null
}

async function getJson(url) {
  try {
    const res = await fetch(url, {
      headers: { Accept: 'application/json' },
      signal: AbortSignal.timeout(15000),
    })
    if (res.status === 429) {
      const retryAfter = parseInt(res.headers.get('retry-after') || '10', 10)
      console.warn(`MetadataFetcher: rate limited, retrying after ${retryAfter}s`)
      await new Promise(r => setTimeout(r, retryAfter * 1000))
      return getJson(url) // single retry
    }
    if (!res.ok) return null
    return await res.json()
  } catch (e) {
    console.warn(`MetadataFetcher: HTTP failed for ${url} — ${e.message}`)
    return null
  }
}

async function search(query) {
  const data = await getJson(`${BASE_URL}/singlesearch/shows?q=${encodeURIComponent(query)}&embed=episodes`)
  if (!data || typeof data !== 'object' || !data.id) return null
  return data
}

// Direct IMDb-id lookup. Returns the full show payload with episodes embedded,
// or null if no match. Best when the id was extracted from a [imdbid-tt…]
// filename marker or NFO sidecar.
async function fetchByImdbId(imdbId) {
  if (!imdbId) return null
  const lookup = await getJson(`${BASE_URL}/lookup/shows?imdb=${encodeURIComponent(imdbId)}`)
  if (!lookup || !lookup.id) return null
  const data = await getJson(`${BASE_URL}/shows/${lookup.id}?embed=episodes`)
  if (!data || !data.id) return null
  return data
}

async function fetchForShow(showId, opts = {}) {
  const s = db.shows.findById(showId)
  if (!s) return false

  // Prefer direct IMDb lookup when known — single exact match, no
  // search fuzziness. Caller (media-scanner.js) seeds opts.imdbId from
  // NFO sidecars and [imdbid-tt…] filename markers.
  let data = null
  if (opts.imdbId) {
    data = await fetchByImdbId(opts.imdbId)
  }
  data ||= await search(s.name)
  if (!data) return false

  // Update show metadata
  const posterUrl = data.image?.original || data.image?.medium || null
  const summary = stripHtml(data.summary)

  db.shows.update(showId, {
    tvmaze_id: data.id,
    poster_url: posterUrl,
    description: summary,
    genres: Array.isArray(data.genres) ? data.genres.join(', ') : null,
    rating: data.rating?.average || null,
    premiered: data.premiered || null,
    status: data.status || null,
    imdb_id: data.externals?.imdb || null,
  })

  // Update episode metadata
  const apiEpisodes = data._embedded?.episodes || []
  if (apiEpisodes.length > 0) {
    const apiLookup = {}
    for (const ep of apiEpisodes) {
      if (ep.season == null || ep.number == null) continue
      const code = `S${String(ep.season).padStart(2, '0')}E${String(ep.number).padStart(2, '0')}`
      apiLookup[code] = ep
    }

    const localEpisodes = db.episodes.forShow(showId)
    let matched = 0
    for (const episode of localEpisodes) {
      const apiEp = apiLookup[episode.code]
      if (!apiEp) continue

      const attrs = {}
      const epSummary = stripHtml(apiEp.summary)
      if (epSummary) attrs.description = epSummary
      if (apiEp.airdate) attrs.air_date = apiEp.airdate
      if (apiEp.runtime) attrs.runtime = apiEp.runtime
      if (apiEp.id) attrs.tvmaze_id = apiEp.id

      if (Object.keys(attrs).length > 0) {
        db.episodes.updateMetadata(episode.id, attrs)
        matched++
      }
    }
    console.log(`MetadataFetcher: matched ${matched}/${localEpisodes.length} episodes with TVMaze data`)
  }

  console.log(`MetadataFetcher: updated show '${s.name}' (TVMaze ID: ${data.id})`)
  return true
}

module.exports = { fetchForShow, fetchByImdbId, search }
